docs/LibraryEvolution.rst

```diff
@@ -145,14 +145,24 @@ The following changes are permitted:
   body, not the labels that are part of the function's full name).
 - Reordering generic requirements (but not the generic parameters themselves).
 - Adding a default argument expression to a parameter.
-- Changing or removing a default argument is a `binary-compatible
-  source-breaking change`.
+- Adding, changing, reordering, or removing property wrappers that either are 
+  implementation-detail (``@ImplementationDetail``) or in a composition where
+  the outermost wrapper is implementation-detail
+  (``@ImplementationDetail @MyWrapper``).
+- Changing or removing a default argument is a `binary-compatible source-breaking change`.
+- Adding, changing, or removing arguments in the property-wrapper custom attribute
```

Comment:
Hmm I don't think this bullet point is right - adding or removing an argument in an API-level property wrapper attribute is an ABI and source breaking change. Changing the argument once it's there is fine.

Reply:
Right. I was thinking that the behavior would be similar to defaulted parameters but didn’t consider that wrapper arguments are chosen by the function author.

Reply:
I think the way this is phrased in the proposal is fine. The section in the proposal that you mentioned specifies that this inference is per wrapper-application, not per wrapper type. I think the right way to phrase it here is to say that changing a property wrapper attribute from API to implementation-detail or vice versa is a source and ABI breaking change.

As far as arguments in composition chains - the check should be if any of the attributes have an argument. If the implementation doesn't do that then it's a bug.

Reply:
> I think the way this is phrased in the proposal is fine. The section in the proposal that you mentioned specifies that this inference is per wrapper-application, not per wrapper type.

Agreed. I'm just worried that the wrapper-level API vs implementation-detail distinction may be more confusing with potential future exceptions.

> As far as arguments in composition chains - the check should be if any of the attributes have an argument. If the implementation doesn't do that then it's a bug.

That check guarantees that no wrapper attributes with arguments are allowed to be enclosed in a composition with an API-level outer wrapper, correct?

>  I think the right way to phrase it here is to say that changing a property wrapper attribute from API to implementation-detail or vice versa is a source and ABI breaking change.

Excellent suggestion; I simplified the rule and place it under the "not allowed" section:

---

An ABI-public function may not change an API-level property-wrapper attribute to an implementation-detail one and vice versa, if it is the only wrapper applied to a given parameter or the outermost wrapper in a composition.